File: backend/src/data/repositories/offer-repository.impl.ts
import { OfferRepository } from '../../domain/repositories/offer-repository.interface';
import { Offer } from '../../domain/entities/offer.entity';
import { OfferModel } from '../models/offer.model';
import { injectable } from 'inversify';

@injectable()

export class OfferRepositoryImpl implements OfferRepository {
  async getOffers(): Promise<Offer[]> {
    const offers = await OfferModel.find();
    return offers.map(offer => new Offer(
      offer._id.toString(),
      offer.title,
      offer.description,
      offer.discountPercentage,
      offer.originalPrice,
      offer.discountedPrice
    ));
  }

  async createOffer(offer: Offer): Promise<Offer> {
    const newOffer = new OfferModel(offer);
    const savedOffer = await newOffer.save();
    return new Offer(
      savedOffer._id.toString(),
      savedOffer.title,
      savedOffer.description,
      savedOffer.discountPercentage,
      savedOffer.originalPrice,
      savedOffer.discountedPrice
    );
  }

  async updateOffer(offer: Offer): Promise<Offer> {
    const updatedOffer = await OfferModel.findByIdAndUpdate(offer.id, offer, { new: true });
    if (!updatedOffer) {
      throw new Error('Offer not found');
    }
    return new Offer(
      updatedOffer._id.toString(),
      updatedOffer.title,
      updatedOffer.description,
      updatedOffer.discountPercentage,
      updatedOffer.originalPrice,
      updatedOffer.discountedPrice
    );
  }

  async deleteOffer(offerId: string): Promise<void> {
    const result = await OfferModel.findByIdAndDelete(offerId);
    if (!result) {
      throw new Error('Offer not found');
    }
  }

  async getOfferById(offerId: string): Promise<Offer | null> {
    const offer = await OfferModel
      .findById(offerId);
    if (!offer) {
      return null;
    }
    return new Offer(
      offer._id.toString(),
      offer.title,
      offer.description,
      offer.discountPercentage,
      offer.originalPrice,
      offer.discountedPrice
    );
  }
}

File: mobile_app/lib/data/repositories/offer_repository_impl.dart
import 'package:dartz/dartz.dart';
import 'package:mobile_app/core/network/network_info.dart';
import 'package:mobile_app/data/datasources/offer_local_data_source.dart';
import 'package:mobile_app/data/models/offer_model.dart';
import '../../domain/entities/offer.dart';
import '../../domain/repositories/offer_repository.dart';
import '../datasources/offer_remote_data_source.dart';
import '../../core/error/failures.dart';
import '../../core/error/exceptions.dart';


class OfferRepositoryImpl implements OfferRepository {
  final OfferRemoteDataSource remoteDataSource;
  // final OfferLocalDataSource localDataSource;
  final NetworkInfo networkInfo;


  OfferRepositoryImpl({
    required this.remoteDataSource,
    // required this.localDataSource,
    required this.networkInfo,
  });

  @override
  Future<Either<Failure, List<Offer>>> getOffers() async {
    if (await _isConnected()) {
      try {
        final remoteOffers = await remoteDataSource.getOffers();
        // localDataSource.cacheOffers(remoteOffers);
        return Right(remoteOffers);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      // try {
      //   final localOffers = await localDataSource.getCachedOffers();
      //   return Right(localOffers);
      // } on CacheException {
      //   return Left(CacheFailure());
      // }
      return Left(NetworkFailure());
    }
  }

  @override
  Future<Either<Failure, Offer>> createOffer(Offer offer) async {
    if (await _isConnected()) {
      try {
        // Convert Offer to OfferModel
        final offerModel = OfferModel(
          id: offer.id,
          title: offer.title,
          description: offer.description,
          discountPercentage: offer.discountPercentage,
          originalPrice: offer.originalPrice,
          discountedPrice: offer.discountedPrice,
        );
        
        final remoteOffer = await remoteDataSource.createOffer(offerModel);
        return Right(remoteOffer);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(NetworkFailure());
    }
  }

  @override
  Future<Either<Failure, Offer>> updateOffer(Offer offer) async {
    if (await _isConnected()) {
      try {
        
        final offerModel = OfferModel(
          id: offer.id,
          title: offer.title,
          description: offer.description,
          discountPercentage: offer.discountPercentage,
          originalPrice: offer.originalPrice,
          discountedPrice: offer.discountedPrice,
        );
        final updatedOffer = await remoteDataSource.updateOffer(offerModel);
        return Right(updatedOffer);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(NetworkFailure());
    }
  }

  @override
  Future<Either<Failure, void>> deleteOffer(String offerId) async {
    if (await _isConnected()) {
      try {
        await remoteDataSource.deleteOffer(offerId);
        return Right(null);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(NetworkFailure());
    }
  }

  @override
  Future<Either<Failure, void>> purchaseOffer(Offer offer, String userId) async {
    if (await networkInfo.isConnected) {
      try {
        await remoteDataSource.purchaseOffer(offer, userId);
        return Right(null);
      } on ServerException {
        return Left(ServerFailure());
      } on PurchaseException {
        return Left(PurchaseFailure());
      }
    } else {
      return Left(NetworkFailure());
    }
  }

  

  Future<bool> _isConnected() async {
    return await networkInfo.isConnected;
  }
}